File: java/june15_3.java

import java.io.*;
import java.util.*;

class SteadyTables
{

    private static int mod = 1000000000;//Modulo value

    public static void main(String [] args)throws IOException
    {

        InputStream inputstream = System.in;
        //OutputStream outputstream = System.out;
        Inp_reader sc = new Inp_reader(inputstream);
        //PrintWriter out = new PrintWriter(outputstream);

    //------Input test cases
        int test;
        test = sc.readInt();
        //@shivamflash

    //------Start the test cases
        for(int t = 0; t < test; t++)
        {

        //------Input rows and columns
            int rows, columns;
            rows = sc.readInt();
            columns = sc.readInt();

            int upper_index[] = new int[columns+1];
            int  lower_index = columns-1;

        //------Finding array of upper indexes in nCr.
            for(int i = 0; i <= columns; i++)
            {
                upper_index[i] = i+lower_index;
            }

            long [] possible = nCr(upper_index, lower_index);


        //------Calculation part.

            //Create a shallow copy of the possible array
            long [] copy_possible = new long[columns+1];

            //System.arraycopy(possible,0,copy_possible,0,columns+1);

            copy_possible = possible.clone();

            for(int row = 0; row < rows-1; row++)
            {

                long temp = copy_possible[0];
                //Add all the possibilities
                for(int index = 1; index <= columns; index++)
                {
                    copy_possible[index]  = (copy_possible[index]%mod) + (copy_possible[index-1]%mod);
                    copy_possible[index-1] = temp;
                    temp = (copy_possible[index]%mod) * (possible[index]%mod);
                }

                copy_possible[columns] = temp;
                //Multiply all the possibilities with the end index
                //for( int index = 1; index <= columns; index++)
                //{
                //   copy_possible[index] = (possible[index]%mod) * (copy_possible[index]%mod);
                //}
            }

        //------Calculate the sum.
            long sum=0;
            for(long i : copy_possible)
            {
                sum = (sum%mod)+(i%mod);
            }

        //------Output
            System.out.println(sum%mod);
        }
    }



    public static long[] nCr(int arr [], int r)
    {
        long combi [] = new long[r+2];
        int c = 0;
        long row[] = new long[arr[l-1]+1];
        row[0] = 1;

        for(int i = 0; i <= arr[l-1]; i++)
        {
            for(int j = i; j > 0; j--)
            {
                row[j] = (row[j]%mod) + ((row[j - 1])%mod);
            }

            if(i == arr[c])
            {
                combi[c++] = row[r]%mod;
            }
        }
        return combi;
    }

}

/*
A table is steady when:
  Sum of ith row is less than or equal to (i-1)th row
  Sum of the last row is less than or equal to number of columns.
@shivamflash
    We need to find the possible steady tables for a given r,c
values. The sum of the last row can be anything between c to 0.
If it is c, then one after it may have sum {c-1 .. 0}
For sum of last row equal to (c-1), others = {c-2 .. 0}
@shivamflash
  To find the sum of any row, we need weak compositions of the
number n into c parts. This will give us all possible combinations
to attain this sum in the row.
It is given by formula (n+c-1, c-1).

*/

class Inp_reader
{

    private InputStream stream;
    private int cChar;
    private byte[] buf = new byte[1024];
    private int numChars;

    public Inp_reader(InputStream stream)
    {
        this.stream = stream;
    }

    public int read()
    {
        if (numChars == -1)
        {
            throw new InputMismatchException();
        }

        if (cChar >= numChars)
        {
            cChar = 0;
            try
            {
                numChars = stream.read(buf);
            }
            catch (IOException e)
            {
                throw new InputMismatchException();
            }
            if (numChars <= 0)
            {
                return -1;
            }
        }
        return buf[cChar++];
    }

    public int readInt()
    {
        int c = read();
        while (isSpaceChar(c))
        {
            c = read();
        }
        int sgn = 1;
        if (c == '-')
        {
            sgn = -1;
            c = read();
        }
        int res = 0;
        do
        {
            if (c < '0' || c > '9') {
            throw new InputMismatchException();
        }
            res *= 10;
            res += c - '0';
            c = read();
        } while (!isSpaceChar(c));

        return res * sgn;
    }

    public String readString()
    {
        int c = read();
        while (isSpaceChar(c))
        {
        c = read();
        }
        StringBuffer res = new StringBuffer();

        do
        {
            res.appendCodePoint(c);
        c = read();
        } while (!isSpaceChar(c));

        return res.toString();
    }

    public static boolean isSpaceChar(int c)
    {
        return c == ' ' || c == '\n' || c == '\r' || c == '\t' || c == -1;
    }

    public double readDouble()
    {
        return Double.parseDouble(readString());
    }

    public long readLong()
    {
        return Long.parseLong(readString());
    }

    public float readFloat()
    {
        return Float.parseFloat(readString());
    }
}

File: c/june15_3.c
#include <stdio.h>
#include <string.h>

void printl(long long arr[], int l)
{
    int i = l;
    for(i =0; i<=l; i++){
        if (arr[i] == 0)
            break;
        printf("%d >> %lld|\n", i , arr[i]);
    }
    printf("\n");
}

void printi(int arr[], int l)
{
    int i = l;
    for(i =0; i<=l; i++){
        if (arr[i] == 0)
            break;
        printf("%d >> %d|\n", i , arr[i]);
    }
    printf("\n");
}

long long* nCr(int *arr, int r, long long mod, long long combi[]){
    int l = r+1;
    int c = 0;
    //arr[l-1]+1) ;
    long long row[3999+1] = {0};
    row[0] = 1;

    printf("upper_index\n");
    printi(arr,1212);
    long long i,j;
    for(i = 0; i <= arr[l]; i++)
    {
        for(j = i; j > 0; j--){
            row[j] = (row[j]%mod) + ((row[j - 1])%mod);
        }

        if(i == arr[c]){
            combi[c] = row[r]%mod;
            c += 1;
        }
    }
    printf("%s\n","possible i.e. weak compositions" );
    printl(combi,1000);
    return combi;
}



void main(void){
    int t;
    scanf("%d",&t);
    while(t--){

        int rows, columns, loop;
        scanf("%d %d",&rows,&columns);

        int upper_index[2000+1];
        int lower_index = columns - 1;
        long long mod = 1000000000;

        for(loop = 0; loop <= columns; loop ++){
            upper_index[loop] = loop + lower_index;
            printf(" upper_index :: %d>%d>>%d\n", loop, lower_index, loop+lower_index);
        }


        long long possible[2000+1] = {0};
        nCr(upper_index, lower_index, mod, possible);


        long long copy_possible[2000+1] = {0};
        for(loop = 0; loop <= columns; loop++){
                copy_possible[loop] = possible[loop];
        }

        printf("Copy of compositions\n");
        printl(copy_possible, 100);

        int row;
        for(row = 0; row < rows-1; row++){
            //Add all the possibilities
            long temp = possible[0];
            for(loop = 1; loop <= columns; loop++){
                copy_possible[loop]  = (copy_possible[loop]%mod) + (copy_possible[loop-1]%mod);
                copy_possible[loop-1] = temp;
                printf("elements \n %lld\n",(copy_possible[loop-1]%mod));
                temp = (copy_possible[loop]%mod) * (possible[loop]%mod);
                //copy_possible[loop]  = (copy_possible[loop]%mod) + (copy_possible[loop-1]%mod);
            }
            copy_possible[columns] = temp;

            /*Multiply all the possibilities with the end loop
            for(loop = 1; loop <= columns; loop++)
            {
                copy_possible[loop] = (possible[loop]%mod) * (copy_possible[loop]%mod);
            }*/

            printf("Ans array %d >> ", row);
            printl(copy_possible, 1000);
        }



        long long sum=0;
        for(loop = 0; loop <= columns; loop++){
            if(copy_possible[loop] == 0)
                break;
            sum = (sum%mod)+(copy_possible[loop]%mod);
        }

        printf("%lld\n",(sum%mod));
    }

}
